### ISSUE TITLE
Proposal: boundaries redesign

### ISSUE THREAD

**Issue by sasa1977** [sasa1977/boundaries#5]
Based on the [feedback gathered on Elixir Forum](https://elixirforum.com/t/boundaries-enforcing-boundaries-in-elixir-projects) and some issues reported here (#2 and #3), I'd like to propose the following changes to the library:

1. Move to a decentralized approach where each boundary is described in its root module.
2. Support unclassified modules

### Usage

Currently, boundaries are defined in a separate `boundaries.exs` file:

```elixir
[
  {MySystem, exports: [User]},
  {MySystemWeb, exports: [Endpoint], deps: [MySystem]}
]
```

In a decentralized approach, each boundary is defined in its root module:

```elixir
defmodule MySystem do
  use Boundary, exports: [User]

  # ...
end

defmodule MySystemWeb do
  use Boundary,
    exports: [Endpoint],
    deps: [MySystem]

  # ...
end
```

Note that with this approach, developers will sometimes need to add a module for the sole purpose of defining a boundary.

Internally, the injected code will be along the lines of:

```elixir
Module.register_attribute(__MODULE__, Boundary, persist: true)
Module.put_attribute(__MODULE__, Boundary, boundary_opts)
```

In particular, no function or macro will be injected into the caller module.

As a result, we lose a central place where one can inspect all the boundaries. However, based on the stored metadata, it's straightforward to add support for that via helper functions such as `Boundary.all/0` and possibly mix tasks which are implemented on top of those functions.

### Motivation

Personally, I think that placing a boundary specification close to its implementation makes more sense, and scales much better with respect to the project size. In a large codebase with many boundaries defined, managing a single large file might become much more difficult.

Another nice benefit is that a boundary name is now auto-inferred from the owner module, which means that if the module is renamed, the boundary name is automatically "synchronized".

Most importantly, a decentralized definition will make it possible to elegantly deal with the protocol implementation, which is a problem that is currently not handled properly.

Consider the following module:

```elixir
defmodule MySystem.SomeModule do
  defimpl Enumerable do
    # ...
  end
end
```

This code will generate the module `Enumerable.MySystem.SomeModule`, which is in a completely different namespace. In the current design, you need to introduce another boundary which follows the same rules as `MySystem.SomeModule`.

In theory, we could deduce that the impl module implements a protocol for `MySystem.SomeModule`, and automatically place the module into the same boundary as `MySystem.SomeModule`, but that decision is not always the right one.

Consider the following situation:

```elixir
# in the interface layer
defimpl SomeLib.Encoder, for: MySystem.SomeModule do ... end
```

In this case, the protocol implementation of the encoder is done for the purposes of the interface layer (i.e. `MySystemWeb`). Thus, in this case, the rules for the protocol implementation should not be the same as the rules for the domain module.

A decentralized approach can help us have an explicit interface to solve this situation:

```elixir
defmodule MySystem.SomeModule do
  defimpl Enumerable do
    # Make this module a part of the `MySystem` boundary
    use Boundary.ProtocolImpl, boundary: MySystem

    # Alternatively, if `boundary` is not specified, the impl belongs to the same
    # boundary as the implementation target (`MySystem.SomeModule`).
    use Boundary.ProtocolImpl
  end
end

# interface layer
defimpl SomeLib.Encoder do
  # Make this module a part of the `MySystemWeb` boundary
  use Boundary.ProtocolImpl, boundary: MySystemWeb
end
```

This feature would work only for modules which are implementing some protocol. Boundary membership of all other modules is determined from the module name, and can't be overridden.

If `use Boundary.ProtocolImpl` is not provided in a protocol impl, the module will be unclassified.

Finally, while not a top priority, a decentralized approach makes it possible to elegantly define internal boundaries (boundaries within boundaries):

```elixir
defmodule MySystem do
  use Boundary, ...
end

defmodule MySystem.Accounts do
  use Boundary, exports: [User, ...]
end

defmodule MySystem.Admin do
  use Boundary, deps: [Accounts, ...]
end
```

In this example, we define two internal boundaries in `MySystem`: `Accounts` and `Admin`, allowing function calls only from `Admin` into `Accounts`.

## Unclassified modules

Currently, all modules must included in some boundary. As it turns out, there are various scenarios where this requirement is too strict.

For example, introducing boundaries in a legacy project might require a large amount of work which can't be done in a single chunk. Thus, a team might instead decide to gradually introduce boundaries to the project. In such cases, we shouldn't report errors on unclassified modules. Omitting such warning would allow the team to detect violations for the currently established boundaries.

In addition, there is a problems of modules which are generated by some macro or the compiler. A simple example of this is using `@derive`. Consider the following code:

```elixir
defmodule MySystem.User do
  @derive Jason.Encoder
end
```

As a result of this, the module `Jason.Encoder.MySystem.User` will be created. This module shouldn't really be considered by boundaries at all, because the author doesn't have any control over the module implementation.

To tackle these problems, I propose the following solutions:

1. By default, we still emit warnings for unclassified modules.
2. An opt-out compiler switch is introduced (e.g. `--ignore-unclassified-modules`).
3. We support an ignore list, a configuration which allows devs to specify namespaces which should never be analyzed by boundaries (e.g. `Jason.Encoder`).

These properties should allow us to tackle all possible scenarios. A new project can start with the default behavior, using the ignore list to deal with `@derive` and other kinds of auto-generated modules. A legacy project can start with `--ignore-unclassified-modules` until they are able to completely cover the boundaries.

**Comment by tcoopman** [comment]
Didn't read everything in detail, but this sounds fine. One thing I do like about the centralized solution is that you can *see* all boundaries in one go. So it might be nice to add some tooling to print out the boundaries of a decentralized solution.

**Comment by sasa1977** [comment]
> One thing I do like about the centralized solution is that you can _see_ all boundaries in one go. So it might be nice to add some tooling to print out the boundaries of a decentralized solution.

Yeah, something will definitely be available. The way I currently envision it, there will be an API function, e.g. `Boundary.all` which would return a structure similar to the current shape of `boundaries.exs`. You could invoke that function from e.g. `iex` session or via `mix run`. On top of that function, we can then build more convenient mix tasks (but these tasks are not an immediate prio). The same function would likely be internaly used by the compiler too.

**Comment by josevalim** [comment]
Maybe protocol implementations should generally have no boundaries? They are not meant to be called directly anyway.

**Comment by sasa1977** [comment]
> Maybe protocol implementations should generally have no boundaries? They are not meant to be called directly anyway.

@josevalim Yeah, but boundaries are also used to restrain what can some piece of code invoke. So, in the example:

```elixir
defmodule MySystem.SomeModule do
  defimpl Enumerable do
    # ...
  end
end
```

I might want to make sure that `Enumerable` impl isn't e.g. invoking something from `MySystemWeb`. To do that, I need to place the `Enumerable` impl into some boundary (e.g. `MySystem`). 

Though, now that you mention it, I typically strive to implement a protocol by making the impl immediately delegate to the struct module, and then optionally adapting the result to the protocol contract, if needed. In other words, I try to keep the implementation thin, so perhaps then enforcing boundaries for protocols isn't all that important. 

WDYT?

**Comment by josevalim** [comment]
> @josevalim Yeah, but boundaries are also used to restrain what can some piece of code invoke. 

Ah, good point. Implementations are meant to be glue code and they can live either close to the domain or close to where they are used, so it is tricky to say. That's why I think "nothing can call it but it can call anything" is a good default that can be overridden by calling `use Boundary`?

**Comment by sasa1977** [comment]
> That's why I think "nothing can call it but it can call anything" is a good default that can be overridden by calling `use Boundary`?

@josevalim  If I understand you right, I believe that I proposed something along those lines.

So in the current proposal, protocol impl by default isn't classified, which means it can call anything. Otherwise, you can explicitly classify it through `use`. 

One nuance is that I propose a special `use Boundary.ProtocolImpl`. The main motivation for this is because I don't think we should allow reclassification of standard modules. So you can only explicitly set boundaries for the protocol impl. Perhaps, `use Boundary, boundary: SomeBoundary` would suffice, if we warn/error if overriding is done on a plain module.

I guess my main question is should we allow reclassification of plain modules? Currently, I'm not in favour of that, but I'm open to counter-arguments :-)

**Comment by josevalim** [comment]
> So in the current proposal, protocol impl by default isn't classified, which means it can call anything. Otherwise, you can explicitly classify it through use.

I see, so it seems we are agreement. I got confused because of this line:

> By default, we still emit warnings for unclassified modules.

Which means implementations would warn. I am saying they don't need to warn by default but it could be overridden. I guess this means I am proposing reclassification?

**Comment by sasa1977** [comment]
> I got confused because of this line:
> 
> > By default, we still emit warnings for unclassified modules.
> 
> Which means implementations would warn. I am saying they don't need to warn by default but it could be overridden. 

Good catch! There's definitely some conflict here. 

How about:

1. We warn by default on unclassified plain modules
2. We don't warn by default for unclassified protocol impl modules.

?

> I guess this means I am proposing reclassification?

Maybe the "reclassification" is not the right term. What I really mean by this is that you can manually assign a boundary to the module. As a result the boundary is not derived from the module name, so you could place e.g. module `Foo` into the boundary `Bar`. Obviously, we need to allow this for protocol impls, but I wouldn't want to allow it for plain modules.

WDYT?

**Comment by josevalim** [comment]
Agreed on all counts.

**Comment by sasa1977** [comment]
Thanks! I'll adapt the proposal, once I've (hopefully) collected some more feedback.

**Comment by hauleth** [comment]
My only concern is whether we should treat boundaries as a documentation or not. If we should treat them as one then maybe reusing `@moduledoc` would be a better idea as (in theory) we could then also present that piece of information in documentation (for example in case of generating documentation for application) or hide it in general (when generating documentation for library). We could even decide about whether given page should end in documentation basing on the boundary name.

**Comment by sasa1977** [comment]
> My only concern is whether we should treat boundaries as a documentation

This is a very interesting idea. However, I'd formulate it somewhat differently.
IMO, boundaries are not documentation, but instead they are boundaries :-) - groups of modules, which you can think of as internal packages within the same project.

You do raise a good point about documentation though. We could use boundaries as a source for enriching the generated docs. A couple of things which immediately come to mind:

- generate a doc page for each boundary, on which boundary's modules, deps, and exports are listed
- include a link to the boundary page for each module
- make the left side navigation pane boundary-oriented (you see the list of boundaries, which you can expand to drill in and see the modules)

TBH I'm not super familiar with ex_doc, so not sure to what extent can this be implemented with the current version. But if boundaries take off, we could definitely try to work with ex_doc team to get better integration with docs. But that's currently not a top priority :-)

In any case, I don't think of boundaries as documentation, so I'm currently not inclined to use `moduledoc` for defining them.

**Comment by sasa1977** [comment]
@josevalim Since I have you here, I have one related question. 

Given what we've discussed, we need a way of figuring out whether a module implements a protocol. From what I can tell, there's no official way to do this, but it's possible by looking into the undocumented `protocol_impl` module attribute. 

Would Elixir team be open to adding the public support for this, for example by introducing the `Protocol.imp/1` function with the following spec:

```elixir
@spec impl(module) :: %{protocol: module, for: module} | nil
```

?

Of course, I'd be happy to submit the PR :-)

**Comment by josevalim** [comment]
@sasa1977 I believe we have inflection functions such as `__impl__` which you can you use to detect implementations.

**Comment by sasa1977** [comment]
> @sasa1977 I believe we have inflection functions such as `__impl__` which you can you use to detect implementations.

@josevalim Yeah, [it exists](https://github.com/elixir-lang/elixir/blob/d6d2e8220d78ec0176b0a5b61eb10dd6cd3102bb/lib/elixir/lib/protocol.ex#L814-L820), thanks!. It's worth noting that this function is also undocumented. I don't mind using an undocumented feature, though I would be a bit happier if it was a part of the official API.

**Comment by josevalim** [comment]
Yes, a PR to document it would be welcome. We already have a section on `Protocol` module docs about reflection, let's include the implementation reflection in there as well.

**Comment by sasa1977** [comment]
Thanks, I'll prep a PR.

**Comment by tomekowal** [comment]
@sasa1977 Would it be possible to use a mix of centralised and decentralised approach? I agree that in most cases, we will use per module definitions but they might get repetitive quickly.
I've recently seen a talk by @studzien about Hexagonal architecture https://youtu.be/XGeK9q6yjsg?t=921

He proposes that in each context we should have a separate module for service, data access and model layer where:
- Model does not depend on anything
- DA layer depends only on model
- Service depends on both Model and DA

I immediately thought about your project. It would be awesome to specify global rules like:

```
[
  MyApp.{AContext}.Service.*, deps: [MyApp.{AContext}.DA.*, MyApp.{AContext}.Model.*]`,
  MyApp.{AContext}.DA.*, deps: [MyApp.{AContext}.Model.*]`
]
```
This way, we could enforce an architecture for entire project in a couple of lines of code and still overwrite it when necessary with `use Boundarires` in the module.

**Comment by sasa1977** [comment]
I've started the work on this, implementing the first version of the decentralized approach. You can track the progress in #6, and provide your feedback there.

**Comment by sasa1977** [comment]
> Would it be possible to use a mix of centralised and decentralised approach? I agree that in most cases, we will use per module definitions but they might get repetitive quickly.
> I've recently seen a talk by @studzien about Hexagonal architecture https://youtu.be/XGeK9q6yjsg?t=921
> 
> He proposes that in each context we should have a separate module for service, data access and model layer where:
> 
>     * Model does not depend on anything
> 
>     * DA layer depends only on model
> 
>     * Service depends on both Model and DA
> 
> 
> I immediately thought about your project. It would be awesome to specify global rules like:
> 
> ```elixir
> [
>   MyApp.{AContext}.Service.*, deps: [MyApp.{AContext}.DA.*, MyApp.{AContext}.Model.*]`,
>   MyApp.{AContext}.DA.*, deps: [MyApp.{AContext}.Model.*]`
> ]
> ```
> 
> This way, we could enforce an architecture for entire project in a couple of lines of code and still overwrite it when necessary with `use Boundarires` in the module.

@tomekowal This is an interesting problem! Here we have the situation where we want to enforce the same constraints in different contexts. So, I presume that the motivation behind wildcards is to somehow allow us to define these constraints once, and have them applied in all target boundaries.

If that's the case, then it's certainly worth thinking about possible solutions, such as the one you proposed. 

Another option which comes to mind is to leave it to devs define these rules through the macro system in their own project. So something like:

```elixir
defmodule MySystem.AContext.SomeModel do
  use MySystem.Model # internally this sets up desired boundary rules for a context model
end
```

The benefit of this is that it doesn't require extra features in boundaries, while leaving the complete freedom to the devs to devise whichever rules they want. 

Not saying that this is better, but just mentioning it as a possible alternative.

In any case, this is not a priority for the moment, so we can spend some time thinking about it, and collecting input from people using the library.

**Comment by sasa1977** [comment]
I added the first draft for supporting ignoring some modules from the boundary check to #6. It's still a bit rough around the edges, but I'm curious to get some feedback on the general approach. Do you think it makes sense? Can you see some possible issues?

TL;DR is that ignoring can be done by marking the boundary as ignored:

```elixir
defmodule SomeBoundary do
  use Boundary, ignore?: true
end
```

For more details, see [here](https://github.com/sasa1977/boundaries/blob/1032db761bcca1da9a6aa8705f889ce12bd9c848/lib/boundary.ex#L126-L185).

**Comment by sasa1977** [comment]
The first draft of all proposed changes is now implemented in PR #6. Note that some parts (most notably ignoring checks) are implemented in a somewhat different way. The relevant links are included in the PR description. I'll leave the PR sit for awhile in hope of getting some feedback.

**Comment by balduncle** [comment]
Hello.  I'm trying out the changes on the `decentralized-approach` branch and have some questions.  Note that all of this is happening in an umbrella project but I'm running my commands within one app in the apps directory, in case that matters.

## Question 1

Does Boundary support custom mix tasks, which per the mix docs are defined in the `Mix.Tasks` namespace?  I want to list a particular module in my project as a dep for the custom task as follows.

```
defmodule Mix.Tasks.MyCustomTask do
  use Boundary, deps: [MySystem], exports: []
  # ...
end
```

But when I do I get a warning like the following.

```
warning: unknown boundary MySystem is listed as a dependency
```

I assume this is because Boundary is trying to find `MySystem` underneath `Mix.Tasks`?  I can avoid the problem above by adding `use Boundary, ignore?: true` to the custom task module.

## Question 2

When I run `mix clean` and `mix compile`, the first time I compile I get legitimate warnings about other missing boundaries.  Generally speaking, I see that warnings are emitted during compilation but it seems that the code is failing to even compile.  As a result I'm running into problems related to modules not being available on subsequent file edits / compiles.

```
warning: function MySystem.SubModule.foo/2 is undefined (module MySystem.Submodule is not available)
...
```

Is this expected?

Thanks for your help.

**Comment by balduncle** [comment]
Never mind on question 1 above.  I forgot that the specific module my custom mix task depended on needed to be exported, since obviously my mix task isn't in my app's namespace.  So only question 2 remains above.  Thanks.

**Comment by sasa1977** [comment]
I'm not quite sure I understand the problem 2. What do you mean by "the code is failing to even compile"? Could you provide a simple repo which demonstrates the issue and the steps to reproduce?

**Comment by balduncle** [comment]
Saša, I am unable to reproduce this problem now. Please forgive me for rushing a report.  I assume it is just me being stupid, but the project I was testing this in is somewhat large and has some interesting cyclic dependencies.  In any event I'll let you know if I can reproduce this reliably.

I have been able to reproduce a different problem in https://github.com/balduncle/my_system_umbrella.  In this case I have a simple Phoenix umbrella app with the Boundary definitions you provide for a typical Phoenix project.  I'm making the assumption that this is supposed to work for umbrella projects as well and assume I might be the first person testing this.  The issue is below.

```
$ mix compile

==> my_system_web
warning: unknown boundary MySystem is listed as a dependency
  /Users/justinjohnson/Code/my_system_umbrella/apps/my_system_web/lib/my_system_web.ex:19

warning: unknown boundary MySystem is listed as a dependency
  /Users/justinjohnson/Code/my_system_umbrella/apps/my_system_web/lib/my_system_web/application.ex:5
```

Am I doing something wrong?  Are boundary definitions not accessible across the umbrella apps?

Thanks again for the library.  It's already been helpful in cleaning up the larger project I maintain.

**Comment by sasa1977** [comment]
>  Are boundary definitions not accessible across the umbrella apps?

Yeah, this is currently not supported. In your setup, `:my_system` OTP app is an external dependency of `:my_system_web`, and you can't reference boundaries of one app in another app. This is something I'm thinking of maybe supporting, but currently it doesn't have a high prio. 

So for the foreseeable time, the scope of boundaries is a single OTP app. Note that this can work in umbrellas too (as your example demonstrates), but you can't cross-reference boundaries between different apps.

**Comment by balduncle** [comment]
Okay, thanks for the clarification.
